src/expr/src/relation/mod.rs

```diff
@@ -1105,104 +1409,151 @@ impl MirRelationExpr {
+    }
+
+    /// Pre-order mutable infallible `MirRelationExpr` visitor for `expr`.
+    fn visit_mut_pre<F>(&self, expr: &mut MirRelationExpr, f: &mut F)
+    where
+        F: FnMut(&mut MirRelationExpr),
+    {
+        f(expr);
+        self.visit1_mut(expr, |e| self.visit_mut_pre(e, f))
+    }
+
+    /// A generalization of [`Self::visit_pre`] and [`Self::visit`].
+    ///
+    /// The function `pre` runs on a `MirRelationExpr` before it runs on any of the
+    /// child `MirRelationExpr`s. The function `post` runs on child `MirRelationExpr`s
+    /// first before the parent.
+    ///
+    /// Optionally, `pre` can return which child `MirRelationExpr`s, if any, should be
+    /// visited (default is to visit all children).
+    fn visit_pre_post<F1, F2>(&self, expr: &MirRelationExpr, pre: &mut F1, post: &mut F2)
+    where
+        F1: FnMut(&MirRelationExpr) -> Option<Vec<&MirRelationExpr>>,
+        F2: FnMut(&MirRelationExpr),
+    {
+        if let Some(to_visit) = pre(expr) {
             for e in to_visit {
-                e.visit_pre_post(pre, post);
+                self.visit_pre_post(e, pre, post);
             }
         } else {
-            self.visit1(|e| e.visit_pre_post(pre, post));
+            self.visit1(expr, |e| self.visit_pre_post(e, pre, post));
         }
-        post(self);
+        post(expr);
     }
 
-    /// Fallible visitor for the [`MirScalarExpr`]s in the relation expression.
-    /// Note that this does not recurse into the `MirScalarExpr`s themselves.
-    pub fn try_visit_scalars_mut<F, E>(&mut self, f: &mut F) -> Result<(), E>
+    /// Fallible visitor for the [`MirScalarExpr`]s directly owned by this relation expression.
+    ///
+    /// The `f` visitor should not recursively descend into owned [`MirRelationExpr`]s.
+    fn try_visit_scalars_mut1<F, E>(&self, expr: &mut MirRelationExpr, f: &mut F) -> Result<(), E>
     where
         F: FnMut(&mut MirScalarExpr) -> Result<(), E>,
     {
         // Match written out explicitly to reduce the possibility of adding a
         // new field with a `MirScalarExpr` within and forgetting to account for it
         // here.
-        self.try_visit_mut(&mut |e| e.try_visit_scalars_mut1(f))
-    }
-
-    /// Fallible visitor for the [`MirScalarExpr`]s directly owned by this relation expression.
-    /// This does not recursively descend into owned [`MirRelationExpr`]s.
-    pub fn try_visit_scalars_mut1<F, E>(&mut self, f: &mut F) -> Result<(), E>
-    where
-        F: FnMut(&mut MirScalarExpr) -> Result<(), E>,
-    {
-        match self {
-            MirRelationExpr::Map { scalars, input: _ }
-            | MirRelationExpr::Filter {
-                predicates: scalars,
-                input: _,
-            } => {
+        match expr {
+            MirRelationExpr::Map { scalars, input: _ } => {
                 for s in scalars {
                     f(s)?;
                 }
                 Ok(())
             }
+            MirRelationExpr::Filter {
```

Comment:
Why do we need to unroll these?

Reply:
Personal preference, and also trying to make the code more readable and aligned between the `Scalar` and `Relation` implementations. 

Is there a measurable performance benefit of packing them in a single match arm?

Reply:
No, there shouldn't be any perf difference. I guess I tend to prefer methods that fit entirely in the screen without having to scroll, so I tend to prefer compacting `match` branches if possible, but I'm fine if the general preference is the opposite.

Reply:
I changed it only here because the body of the `*visit*` method is strictly derived from the shape of the `MirRelationExpr` enum, and I wanted to make it more obvious (other recursive arms in `transform` are left intact in #9293).